# Exploring Hackers News Posts
### Ask HN vs Show HN

In this project, we'll compare two different types of posts from Hacker News, a popular site where technology related stories (or 'posts') are voted and commented upon. The two types of posts we'll explore begin with either Ask HN or Show HN.

Users submit Ask HN posts to ask the Hacker News community a specific question, such as "What is the best online course you've ever taken?" Likewise, users submit Show HN posts to show the Hacker News community a project, product, or just generally something interesting.

We'll specifically compare these two types of posts to determine the following:

- Do Ask HN or Show HN receive more comments on average?
- Do posts created at a certain time receive more comments on average?

It should be noted that the data set we're working with was reduced from almost 300,000 rows to approximately 20,000 rows by removing all submissions that did not receive any comments, and then randomly sampling from the remaining submissions..



## Introduction

First, we'll read in the data and remove the headers.


In [1]:
from csv import reader
opened_file = open("hacker_news.csv")
read_file = reader(opened_file)
hn = list(read_file)
print(hn[:5])



[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]



## Removing Headers from a List of Lists


In [2]:
headers = hn[0]
hn = hn[1:]

print(headers)
print("\n")
print(hn[:5])

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01'], ['10301696', 'Note by Note: The Making of Steinway L1037 (2007)', 'http://www.nytimes.com/2007/11/07/movies/07stein.html?_r=0', '8', '2', 'walterbell', '9/30/2015 4:12']]


## Extracting Ask HN and Show HN Posts

First, we'll identify posts that begin with either Ask HN or Show HN and separate the data for those two types of posts into different lists. Separating the data makes it easier to analyze in the following steps.

In [3]:
# Identify posts that begin with either `Ask HN` or `Show HN` and separate the data into different lists.
ask_posts = []
show_posts = []
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    
    if title.startswith("ask hn"):
        ask_posts.append(row)
    elif title.startswith("show hn"):
        show_posts.append(row)
    else: 
        other_posts.append(row)

print(len(ask_posts), len(show_posts), len(other_posts))  


1744 1162 17194


In [4]:
print(ask_posts[:5])
print("\n")
print(show_posts[:5])

[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 

## Calculating the Average Number of Comments for Ask HN and Show HN Posts

Now that we separated ask posts and show posts into different lists, we'll calculate the average number of comments each type of post receives.

In [5]:
total_ask_comments = 0
for row in ask_posts:
    num_comments = int(row[4])
    total_ask_comments += num_comments
    
avg_ask_comments = total_ask_comments / len(ask_posts)
print(avg_ask_comments)
    

14.038417431192661


In [6]:
total_show_comments = 0
for row in show_posts:
    num_comments = int(row[4])
    total_show_comments += num_comments

avg_show_comments = total_show_comments / len(show_posts)
print(avg_show_comments)

10.31669535283993


On average, ask posts in our sample receive approximately 14 comments, whereas show posts receive approximately 10. Since ask posts are more likely to receive comments, we'll focus our remaining analysis just on these posts.


## Finding the Amount of Ask Posts and Comments by Hour Created

Next, we'll determine if we can maximize the amount of comments an ask post receives by creating it at a certain time. First, we'll find the amount of ask posts created during each hour of day, along with the number of comments those posts received. Then, we'll calculate the average amount of comments ask posts created at each hour of the day receive.

In [7]:
# Calculate the amount of ask posts created during each hour of day and the number of comments received.
import datetime as dt

result_list = []

for row in ask_posts:
    created_at = row[6]
    num_comments = int(row[4])

    result_list.append([created_at, num_comments])

counts_by_hour = {}
comments_by_hour = {}

for row in result_list:
    hour = row[0]
    comments = row[1]
    dt_hour = dt.datetime.strptime(hour, "%m/%d/%Y %H:%M")
    dt_format = dt_hour.strftime("%H")
    
    if dt_format not in counts_by_hour:
        counts_by_hour[dt_format] = 1
        comments_by_hour[dt_format] = comments
    else:
        counts_by_hour[dt_format] += 1
        comments_by_hour[dt_format] += comments


print(comments_by_hour)
print("\n")
print(counts_by_hour)

{'03': 421, '11': 641, '20': 1722, '00': 447, '17': 1146, '19': 1188, '08': 492, '04': 337, '12': 687, '09': 251, '16': 1814, '18': 1439, '14': 1416, '10': 793, '15': 4477, '01': 683, '07': 267, '02': 1381, '22': 479, '23': 543, '13': 1253, '06': 397, '21': 1745, '05': 464}


{'03': 54, '11': 58, '20': 80, '00': 55, '17': 100, '19': 110, '08': 48, '04': 47, '12': 73, '09': 45, '16': 108, '18': 109, '14': 107, '10': 59, '15': 116, '01': 60, '07': 34, '02': 58, '22': 71, '23': 68, '13': 85, '06': 44, '21': 109, '05': 46}


## Calculating the Average Number of Comments for Ask HN Posts by Hour

In [13]:
# Calculate the average amount of comments `Ask HN` posts created at each hour of the day receive.

avg_by_hour =[]
for hour in comments_by_hour:
    avg_by_hour.append([hour, (comments_by_hour[hour]/counts_by_hour[hour])])
    
avg_by_hour

[['03', 7.796296296296297],
 ['11', 11.051724137931034],
 ['20', 21.525],
 ['00', 8.127272727272727],
 ['17', 11.46],
 ['19', 10.8],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['12', 9.41095890410959],
 ['09', 5.5777777777777775],
 ['16', 16.796296296296298],
 ['18', 13.20183486238532],
 ['14', 13.233644859813085],
 ['10', 13.440677966101696],
 ['15', 38.5948275862069],
 ['01', 11.383333333333333],
 ['07', 7.852941176470588],
 ['02', 23.810344827586206],
 ['22', 6.746478873239437],
 ['23', 7.985294117647059],
 ['13', 14.741176470588234],
 ['06', 9.022727272727273],
 ['21', 16.009174311926607],
 ['05', 10.08695652173913]]

## Sorting and Printing Values from a List of Lists

In [27]:
swap_avg_by_hour = []
for row in avg_by_hour:
    swap_avg_by_hour.append([row[1],row[0]])
    
print(swap_avg_by_hour)
sorted_swap = sorted(swap_avg_by_hour, reverse=True)

# Sort the values and print the the 5 hours with the highest average comments.
print("\n")
print("Top 5 Hours for Ask Post Comments")


for avg, hr in sorted_swap[:5]:
    print("{0}: {1:.2f} average comments per post".format(
        dt.datetime.strptime(hr, "%H").strftime("%H:%M"), avg))

    
    


[[7.796296296296297, '03'], [11.051724137931034, '11'], [21.525, '20'], [8.127272727272727, '00'], [11.46, '17'], [10.8, '19'], [10.25, '08'], [7.170212765957447, '04'], [9.41095890410959, '12'], [5.5777777777777775, '09'], [16.796296296296298, '16'], [13.20183486238532, '18'], [13.233644859813085, '14'], [13.440677966101696, '10'], [38.5948275862069, '15'], [11.383333333333333, '01'], [7.852941176470588, '07'], [23.810344827586206, '02'], [6.746478873239437, '22'], [7.985294117647059, '23'], [14.741176470588234, '13'], [9.022727272727273, '06'], [16.009174311926607, '21'], [10.08695652173913, '05']]


Top 5 Hours for Ask Post Comments
15:00: 38.59 average comments per post
02:00: 23.81 average comments per post
20:00: 21.52 average comments per post
16:00: 16.80 average comments per post
21:00: 16.01 average comments per post




The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

According to the data set [documentation](https://www.kaggle.com/hacker-news/hacker-news-posts/home), the timezone used is Eastern Time in the US. So, we could also write 15:00 as 3:00 pm est.


## Conclusion

In this project, we analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on our analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.
